src/Utopia/Messaging/Adapters/SMS/GEOSMS.php

```diff
@@ -0,0 +1,289 @@
+        '46' => 'Sweden (+46)',
+        '41' => 'Switzerland (+41)',
+        '963' => 'Syria (+963)',
+        '886' => 'Taiwan (+886)',
+        '7' => 'Tajikstan (+7)',
+        '66' => 'Thailand (+66)',
+        '228' => 'Togo (+228)',
+        '676' => 'Tonga (+676)',
+        '1868' => 'Trinidad & Tobago (+1868)',
+        '216' => 'Tunisia (+216)',
+        '90' => 'Turkey (+90)',
+        '7' => 'Turkmenistan (+7)',
+        '993' => 'Turkmenistan (+993)',
+        '1649' => 'Turks & Caicos Islands (+1649)',
+        '688' => 'Tuvalu (+688)',
+        '256' => 'Uganda (+256)',
+        '380' => 'Ukraine (+380)',
+        '971' => 'United Arab Emirates (+971)',
+        '598' => 'Uruguay (+598)',
+        '7' => 'Uzbekistan (+7)',
+        '678' => 'Vanuatu (+678)',
+        '379' => 'Vatican City (+379)',
+        '58' => 'Venezuela (+58)',
+        '84' => 'Vietnam (+84)',
+        '84' => 'Virgin Islands - British (+1284)',
+        '84' => 'Virgin Islands - US (+1340)',
+        '681' => 'Wallis & Futuna (+681)',
+        '969' => 'Yemen (North)(+969)',
+        '967' => 'Yemen (South)(+967)',
+        '260' => 'Zambia (+260)',
+        '263' => 'Zimbabwe (+263)',
+    ];
+
+    protected $defaultAdapter;
+
+    protected $localAdapters = [];
+
+    protected $maxMessagesPerRequest;
+
+    public function __construct(SMSAdapter $defaultAdapter)
+    {
+        $this->defaultAdapter = $defaultAdapter;
+        $this->maxMessagesPerRequest = $defaultAdapter->getMaxMessagesPerRequest();
+    }
+
+    public function getName(): string
+    {
+        return 'GEOSMS';
+    }
+
+    public function getMaxMessagesPerRequest(): int
+    {
+        return $this->maxMessagesPerRequest;
+    }
+
+    public function setLocal(string $prefix, SMSAdapter $adapter): self
+    {
+        $prefix = preg_replace('/\+/', '', $prefix);
+        $this->localAdapters[$prefix] = $adapter;
+        $this->maxMessagesPerRequest = min($this->maxMessagesPerRequest, $adapter->getMaxMessagesPerRequest());
+
+        return $this;
+    }
+
+    protected function process(SMS $message): string
+    {
+        $adapter = $this->getAdapterForMessage($message);
+
+        return $adapter->send($message);
+    }
+
+    protected function getAdapterForMessage(SMS $message): SMSAdapter
+    {
+        foreach ($message->getTo() as $recipient) {
+            $prefix = $this->extractPrefix($recipient);
+            if (! empty($prefix) && array_key_exists($prefix, $this->localAdapters)) {
+                return $this->localAdapters[$prefix];
+            }
+        }
+
+        return $this->defaultAdapter;
+    }
+
+    protected function extractPrefix(string $phoneNumber): string
+    {
+        $digits = preg_replace('/\D/', '', $phoneNumber);
+        foreach ([3, 2, 1] as $length) {
+            $prefix = substr($digits, 0, $length);
+            if (isset(self::COUNTRY_CODES[$prefix])) {
+                return $prefix;
+            }
+        }
```

Comment:
Does this also account for these scenarios 
1. `0091<number>` and 
2. `+91<number>`

Reply:
00 is an international dialling code (signals you are making an international call) for most of the world but it's actually different in some places. In the US it's 011. 

International dialling codes are mostly obsolete since you can use +

Reply:
Good to know!